crates/sui-storage/src/fullnode_pending_tx_log.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! FullNodePendingTransactionLog is Write Ahead Log uesd in Transaction Orchestrator
+//! for transaction submission processing. It plays two roles:
+//! 1. At one time, a transaction is only processed once.
+//! 2. When Fullnode crashes and restarts, the pending transaction will be loaded and retried.
+
+use crate::mutex_table::MutexTable;
+use crate::write_ahead_log::{MUTEX_TABLE_SHARD_SIZE, MUTEX_TABLE_SIZE};
+use std::path::PathBuf;
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{TrustedTransction, VerifiedTransaction};
+use typed_store::traits::TypedStoreDebug;
+use typed_store::{rocks::DBMap, traits::Map};
+use typed_store_derive::DBMapUtils;
+
+pub type IsFirstRecord = bool;
+
+#[derive(DBMapUtils)]
+struct FullNodePendingTransactionTable {
+    logs: DBMap<TransactionDigest, TrustedTransction>,
+}
+
+pub struct FullNodePendingTransactionLog {
+    pending_transactions: FullNodePendingTransactionTable,
+    mutex_table: MutexTable<TransactionDigest>,
+}
+
+impl FullNodePendingTransactionLog {
+    pub fn new(path: PathBuf) -> Self {
+        let pending_transactions =
+            FullNodePendingTransactionTable::open_tables_read_write(path, None, None);
+        Self {
+            pending_transactions,
+            mutex_table: MutexTable::new(MUTEX_TABLE_SIZE, MUTEX_TABLE_SHARD_SIZE),
+        }
+    }
+
+    pub async fn write_pending_transaction(
+        &self,
+        tx: &VerifiedTransaction,
+    ) -> SuiResult<IsFirstRecord> {
+        let tx_digest = tx.digest();
+        let _guard = self.mutex_table.acquire_lock(*tx_digest).await;
```

Comment:
I don't quite understand why we need a lock guard here?

Reply:
This is to prevent race conditions between the following db read and write. It's possible that two requests (for the same transaction) are in this region and read results are empty, then both of them think they are the first occurrence. Then they would enqueue the transaction twice for processing

Reply:
logs is a DBMap keyed by digest, so it will never enqueue the same transaction "twice"

Reply:
it will enqueue to the queue (to be introduced) in Orchestrator twice.

Reply:
https://github.com/MystenLabs/sui/pull/6710/files#r1046167134